### PR TITLE
#160451499 Password Reset Functionality

### PR DESCRIPTION
**Pull request by slimsolz**
#### What does this PR do?
It sends user reset password URL mail and allows a user to update his password

#### Description of Task to be completed?
- Create a sendMail function to send mail to the user using SendGrid
- Send mail to the user
- Validate the URL
- Allow user update password

#### How should this be manually tested?
- Clone the repo with "git clone https://github.com/andela/phlox-ah.git"
- cd into phlox-ah. cd phlox-ah
- Switch to the feature branch git branch -r then run git checkout feature/160451499/user-reset-password
- Run npm install
- Rename .env_sample file to .env then fill it with your locally created Postgres database details
- If you have PostgreSQL install you can create a database on your command line with createdb your_db_name
- Run npm run migrate-reset to migrate your table into your database
- Run npm run dev to start your server
- Open postman, then visit localhost:3000/api/v1//forget a Post request with the following form entry: email.
- A mail will be sent to your email account click to reset password
- Open postman, then visit localhost:3000/api/v1/reset_password a Put request with the following form entry: password.
- A successful registration completed will be displayed

#### What are the relevant pivotal tracker stories?
#160451499

### CODE REVIEW

**Review by udochukwu — COMMENTED** [review]
LGTM

**Comment by samdiano** [comment]
LGTM

File: server/controllers/userController.js [hunk 1]
```diff
@@ -29,4 +31,79 @@ export default class UserController {
...
+    const { email } = req.body;
+
+    User.findOne({ where: { email } })
+      .then((userFound) => {
+        const { username } = userFound;
+        if (!userFound) {
+          return res.status(400).json({ success: false, message: 'Email address is not registered' });
+        }
+        const resetToken = crypto.randomBytes(16).toString('hex');
+        const options = {
```
Comment (s-akinrele) [inline server/controllers/userController.js:80]: confirm the copy text with your PO ..

File: server/controllers/userController.js [hunk 2]
```diff
@@ -29,4 +31,79 @@ export default class UserController {
...
+  * @param {object} req - The request payload sent to the router
+  * @param {object} res - The response payload sent back from the controller
+  * @returns {object} - status Message and confirms the mail has been sent
+  */
+  static resetPassword(req, res) {
+    const resetToken = req.params.token;
+    const { password } = req.body;
+
+    User.findOne({ where: { resetToken } })
+      .then((userFound) => {
```
Comment (s-akinrele) [inline server/controllers/userController.js]: just use `user`

File: server/helpers/mail.js [hunk 3]
```diff
@@ -0,0 +1,15 @@
+import sgMail from '@sendgrid/mail';
+
+const sendMail = (options) => {
+  sgMail.setApiKey(process.env.SENDGRID_API_KEY);
+  const mailMessage = {
```
Comment (s-akinrele) [inline server/helpers/mail.js]: sync with @onwuvic to use the same helper.

File: server/routes/api/user.js [hunk 4]
```diff
@@ -13,6 +13,8 @@ router.get('/', (req, res, next) => res.json('welcome Author Haven'));
 
 /* User Endpoint */
 router.post('/signup', UserValidations.validateSignup, UserController.create);
+router.post('/forget', UserController.forgetPassword);
```
Comment (s-akinrele) [inline server/routes/api/user.js]: forgotPassword would be more appropriate here ..

**Review by s-akinrele — CHANGES_REQUESTED** [review]
Please update the PR with your test. Looking good overall.

File: server/controllers/userController.js (hunk 1 above)
Comment (onwuvic) [reply server/controllers/userController.js:80]: PO has agreed with the text.

**Review by onwuvic — COMMENTED** [review]
LGTM

**Review by s-akinrele — APPROVED** [review]
GTM

**Pull request merged**